### ISSUE TITLE
Should casting array-of-T to T require a T-to-T cast?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#23371]
After merging #23352, it turns out I broke Arkouda because:

* I had removed the bool-to-bool cast thinking a self-cast was pointless and removed by the compiler anyway
* yet Arkouda was relying on a cast from an array-of-bools to bool (essentially, a promoted no-op cast), and this doesn't resolve without the explicit bool-to-bool self-cast.

This can be seen in other contexts if you define a type `t`, declare an array of `t` and then try to cast the array-of-t to `t`.  If you don't have an explicit t-to-t cast, it won't resolve.

This issue proposes resolving this by introducing a cast from `array-of-t` to `t` that simply returns the array.  In cases where the array is dead, this could even result in a steal of the array.

Here's a proposed implementation:

```chapel
  operator :(in x: [] ?et, type t: et) where t == et {
    return x;
  }
```

And here's a draft PR that adds it, with a test, removing the explicit bool->bool cast: https://github.com/chapel-lang/chapel/pull/23370

**Comment by bradcray** [comment]
> except maybe for writing out different cast cases (some of which are not no-op) symmetrically?

That's my thinking—in a generic context, you may just be casting one thing to another without knowing much about what the types are.  Though I'm not familiar with the Arkouda context in which this is happening, that's what it looked like at a glance:  A very generic array casting context that happened to be doing an array-of-bool to bool cast in some cases.

We could similarly disallow self casts in non-promoted contexts (i.e., I can't write `42: int` or `myVar: myVar.type` because they're pointless), but historically we've permitted them and optimized them away.  It seemed more natural to me to support array-of-T to T casts for symmetry rather than to say that some pointless casts are supported and others are not.

> Additionally, if we did this I would like it to be pragma "last resort" to make sure we never conflict with a more specific user-defined cast, though from my understanding it's unlikely to conflict anyways.

Makes sense.

**Comment by bradcray** [comment]
This was resolved by the merge of #23370.
